src/NumericalAlgorithms/Spectral/SwshInterpolation.cpp

```diff
@@ -0,0 +1,622 @@
+    const gsl::not_null<SpinWeighted<ComplexDataVector, Spin>*> harmonic,
+    const SpinWeighted<ComplexDataVector, Spin>& harmonic_at_l_min,
+    const int m) const noexcept {
+  const auto& clenshaw_factors = cached_clenshaw_factors<Spin>(l_max_);
+  const double a = std::abs(Spin + m);
+  const double b = std::abs(Spin - m);
+  harmonic->data() =
+      clenshaw_factors
+          // clang-tidy: this is the right order of the casts, other orders give
+          // the wrong answer
+          // NOLINTNEXTLINE(misc-misplaced-widening-cast)
+          .harmonic_at_l_min_plus_one_recurrence_prefactors[static_cast<size_t>(
+              m + static_cast<int>(l_max_))] *
+      harmonic_at_l_min.data() *
+      (a + 1.0 + 0.5 * (a + b + 2.0) * (cos_theta_ - 1.0));
+}
+
+template <int Spin>
+void SwshInterpolator::evaluate_swsh_m_recurrence_at_l_min(
+    const gsl::not_null<SpinWeighted<ComplexDataVector, Spin>*> harmonic,
+    const int m) const noexcept {
+  const auto& clenshaw_factors = cached_clenshaw_factors<Spin>(l_max_);
+  harmonic->data() =
+      // this is the right order of the casts, other orders give the wrong
+      // answer
+      // NOLINTNEXTLINE(misc-misplaced-widening-cast)
+      clenshaw_factors.harmonic_m_recurrence_prefactors[static_cast<size_t>(
+          m + static_cast<int>(l_max_))] *
+      (sin_theta_ / 2.0) * harmonic->data();
+  if (m > 0) {
+    harmonic->data() *= (std::complex<double>(1.0, 0.0) * cos_m_phi_[1] +
+                         std::complex<double>(0.0, 1.0) * sin_m_phi_[1]);
+  } else {
+    harmonic->data() *= (std::complex<double>(1.0, 0.0) * cos_m_phi_[1] -
+                         std::complex<double>(0.0, 1.0) * sin_m_phi_[1]);
+  }
+}
+
+template <int Spin>
+void SwshInterpolator::clenshaw_sum(
+    const gsl::not_null<SpinWeighted<ComplexDataVector, Spin>*> interpolation,
+    const SpinWeighted<ComplexDataVector, Spin>& l_min_harmonic,
+    const SpinWeighted<ComplexDataVector, Spin>& l_min_plus_one_harmonic,
+    const SpinWeighted<ComplexModalVector, Spin>& goldberg_modes,
+    const int m) const noexcept {
+  // Since we need various combinations of the three-term recurrence constants
+  // up to two orders higher, we write recurrence results to a cyclic
+  // three-element cache
+  std::array<ComplexDataVector, 3> recurrence_cache;
+  recurrence_cache[2] = ComplexDataVector{interpolation->size(), 0.0};
+  recurrence_cache[1] = ComplexDataVector{interpolation->size(), 0.0};
+  recurrence_cache[0] = ComplexDataVector{interpolation->size(), 0.0};
+  const auto& clenshaw_factors = cached_clenshaw_factors<Spin>(l_max_);
+
+  for (auto l = static_cast<int>(l_max_);
+       l > std::max(std::abs(Spin), std::abs(m)); l--) {
+    // we want to cycle 'backwards' through the cache for the intuition that
+    // we start with high values and go to low ones. However, the modular
+    // arithmetic demands that we have positive offsets, so we exchange l -
+    // l_max for l + 2 l_max
```

Comment:
How about the following comment instead:

'We want to define some cache_offset so that we can index the three elements of recurrence_cache with indices cache_offset%3, (cache_offset+1)%3, and (cache_offset+2)%3, and so that cache_offset decreases by one on each iteration.  The "obvious" way to do this is to choose cache_offset = l - l_max, so that cache_offset starts at zero and then decreases each iteration.  However, this gives negative values of cache_offset, and modular arithmetic gets complicated with negative numbers.  But note that adding any multiple of 3 to the "obvious" value of cache_offset will give identical indexing, and choosing this multiple of 3 large enough (i.e. larger than l_max) guarantees that the new cache_offset is positive for all l.  So we choose to add 3*l_max to the "obvious" value of cache_offset; in other words we define cache_offset = l + 2 l_max.'

In the future, if this sort of thing is done more often, it may be useful to define an iterator with an `operator[]`, so you can say `iter[i]` instead of `(cache_offset+i)%3`.